Add --installroot support

jrohel opened the pull request:
The "--installroot" argument must be used together with "--config", "--noplugins", "--setopt=cachedir=<path>", "--setopt=reposdir=<path>", and "--setopt=varsdir=<path>" arguments.
The release version is detected from the installroot directory. If it is not possible (eg. empty/newly created installroot) then the release version must be supplied as an additional argument "--releasever=<requested_releasever>".

Fixes https://github.com/rpm-software-management/microdnf/issues/19.

Depends on https://github.com/rpm-software-management/libdnf/pull/875

Conan-Kudo commented on dnf/dnf-main.c:
```diff
@@ -40,21 +41,55 @@ process_global_option (const gchar  *option_name,
+           * If /etc/yum.repos.d/ or /etc/repos.d/ subdirectory exists in installroot directory,
+           * it will be used.
+           */
+          const char * const REPOS_CFG_DIR_PATHS[] = {"/etc/yum.repos.d/", "/etc/repos.d/", NULL};
```
Where did `/etc/repos.d` come from? Is that a path we support somewhere?

jrohel replied:
@Conan-Kudo Oh. You are right. I looked to the history. Dnf is looking for these directories "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d". I will fix it.

Conan-Kudo replied:
@jrohel We also need vars parsed, right?

jrohel replied:
@Conan-Kudo The basic vars (eg. releasever) are detected from installroot (if not overided by command line). Loading vars from files and environment is not working in microdnf at all. I have to finish some work for it. But I hope --installroot is usable just now.

Conan-Kudo replied:
Don't we have an API in libdnf to fetch this stuff now? I vaguely recall that DNF now uses that...

jrohel replied:
Yes, we have API for configuring paths to vars directories. It is not fully functional in libdnf yet. It depends on https://github.com/rpm-software-management/libdnf/pull/691 . I must finish it. I'm stopped the work for a while because I am thinking about DNF API compatibility.
Anyway, idea is, that libdnf will read vars from installroot directory by default. So there is not need for setup from microdnf side.

jrohel commented:
We are thinking about change of repositories configuration directory auto detection algorithm in DNF. After clarifying the algorithm I will modify the PR. We want similar behaviour in DNF and microdnf.

Conan-Kudo commented:
@jrohel Still interested in fixing this?

dmach commented:
@Conan-Kudo I believe we still want this. I'll check with @jrohel what's the remaining work and impact. Adding the feature itself is quite straightforward, but we'll most likely bump into issues that other functionality is not prepared for installroot. It will require adding several tests to ci-dnf-stack.

Conan-Kudo commented:
@jrohel Can you rebase this?

Conan-Kudo commented:
@jrohel @dmach For what it's worth, getting microdnf to use the libdnf API to determine repo configuration directories and var configuration directories is the current blocker for putting microdnf in openSUSE. And missing installroot support blocks adding support for microdnf to kiwi.

jrohel commented:
@Conan-Kudo 
I worked on the PR some time ago. But I stopped work on it. Why?

There are some questions about paths. Path to the master configuration ("dnf.conf"), path to the repository directory, to the vars, to the cache, ...
In the DNF is automatic computation of the paths. Eg. if the reposdirectory exists in installroot it is used else repository configuration from host is used. Jaroslav Mracek is not satisfied with the algorithm and we thinking about change in DNF5. He wanted me to stop my work on the PR until we reach an agreement on final algorithm.
I think installing into another root is useful. So, what now? I have one idea. I can create another PR which adds more options to microdnf: --config=,  --setopt=reposdir=, --setopt=varsdir=, --setopt=cachedir=, ... And then finish installroot support. And user will be forced to state the paths explicitly.
Eg:
`microdnf --noplugins --config=/myroot/etc/dnf/dnf.conf --setopt=reposdir=/myroot/etc/yum.repos.d/ --setopt=varsdir=/myroot/etc/dnf/vars --setopt=cachedir=/myroot/var/cache/yum --installroot=/myroot --releasever=31 install something`
I'm not sure If I will have time for it.

**And what do you think about the idea with more options?**
Oh. I see your new comment now. More microdnf options are required anyway.

Conan-Kudo commented:
> In the DNF is automatic computation of the paths. Eg. if the reposdirectory exists in installroot it is used else repository configuration from host is used. Jaroslav Mracek is not satisfied with the algorithm and we thinking about change in DNF5. He wanted me to stop my work on the PR until we reach an agreement on final algorithm.

I think that changing the way the paths are searched (the algorithm for this, if you will) is somewhat immaterial to the fact that there's already repo search APIs in libdnf itself. We should be using them now in both microdnf and PackageKit-DNF so that there's consistency in how the paths are accessed with DNF itself.

Moreover, we're probably going to start working on supporting override files in libdnf in the next couple of months, so that means it's better now to centralize the logic for handling reading configuration data in libdnf rather than quibbling about how the data access should be changed in the future. If we centralize the logic in libdnf, we can change it pretty freely without impacting all the DNF applications (DNF, Micro DNF, PackageKit, etc.).

jrohel commented:
@Conan-Kudo
> For what it's worth, getting microdnf to use the libdnf API to determine repo configuration directories and var configuration directories is the current blocker for putting microdnf in openSUSE.

I had a vacation for two weeks. But during it, I worked a little bit on libdnf and microdnf. So, I created some PRs.

I hope that with PRs below you have all what you need for putting microdnf in openSUSE:
https://github.com/rpm-software-management/libdnf/pull/868
https://github.com/rpm-software-management/microdnf/pull/67
https://github.com/rpm-software-management/microdnf/pull/68
https://github.com/rpm-software-management/microdnf/pull/69
https://github.com/rpm-software-management/microdnf/pull/70

> And missing installroot support blocks adding support for microdnf to kiwi.

I will be working on a new libdnf. In the meantime, I'll see what to do with installroot in microdnf.

Conan-Kudo commented:
@jrohel Thanks for this! I'm going through and reviewing your PRs now. 👍

Conan-Kudo commented:
@jrohel Everything has been merged except for this PR (which needs rebasing and fixing).

jrohel commented:
@Conan-Kudo I rebased and updated the PR. But first must be merged libdnf fix. It fixes path to history database in combination with installroot. https://github.com/rpm-software-management/libdnf/pull/875

Conan-Kudo commented:
So that's merged, now I'll merge this.
